## Problem Statement

### Business Context

The healthcare industry is rapidly evolving, with professionals facing increasing challenges in managing vast volumes of medical data while delivering accurate and timely diagnoses. The need for quick access to comprehensive, reliable, and up-to-date medical knowledge is critical for improving patient outcomes and ensuring informed decision-making in a fast-paced environment.

Healthcare professionals often encounter information overload, struggling to sift through extensive research and data to create accurate diagnoses and treatment plans. This challenge is amplified by the need for efficiency, particularly in emergencies, where time-sensitive decisions are vital. Furthermore, access to trusted, current medical information from renowned manuals and research papers is essential for maintaining high standards of care.

To address these challenges, healthcare centers can focus on integrating systems that streamline access to medical knowledge, provide tools to support quick decision-making, and enhance efficiency. Leveraging centralized knowledge platforms and ensuring healthcare providers have continuous access to reliable resources can significantly improve patient care and operational effectiveness.

**Common Questions to Answer**

**1. Diagnostic Assistance**: "What are the common symptoms and treatments for pulmonary embolism?"

**2. Drug Information**: "Can you provide the trade names of medications used for treating hypertension?"

**3. Treatment Plans**: "What are the first-line options and alternatives for managing rheumatoid arthritis?"

**4. Specialty Knowledge**: "What are the diagnostic steps for suspected endocrine disorders?"

**5. Critical Care Protocols**: "What is the protocol for managing sepsis in a critical care unit?"

### Objective

As an AI specialist, your task is to develop a RAG-based AI solution using renowned medical manuals to address healthcare challenges. The objective is to **understand** issues like information overload, **apply** AI techniques to streamline decision-making, **analyze** its impact on diagnostics and patient outcomes, **evaluate** its potential to standardize care practices, and **create** a functional prototype demonstrating its feasibility and effectiveness.

### Data Description

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

## Installing and Importing Necessary Libraries and Dependencies

In [ ]:
# Installation for GPU llama-cpp-python
# uncomment and run the following code in case GPU is being used
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

# Installation for CPU llama-cpp-python
# uncomment and run the following code in case GPU is not being used
# !CMAKE_ARGS="-DLLAMA_CUBLAS=off" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.4/9.4 MB 123.4 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 214.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 341.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 308.6 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
opencv-python 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 2.4.1 which is incompatible.
opencv-python-headless 4.12.0.88 requires numpy<2.3.0,>=2; python_version >= "3.9", but you have numpy 2.4.1 which is incompatible.
opencv-contrib-python 4.12.0.88 requires numpy<2.3.0,>=2; python_version

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [ ]:
!pip install --no-cache-dir numpy==1.26.4 pandas==2.1.4

In [ ]:
# For installing the libraries & downloading models from HF Hub
!pip install huggingface_hub==0.35.0 pandas==2.2.2 tiktoken==0.11.0 pymupdf==1.26.4 langchain==0.3.27 langchain-community==0.3.29 chromadb==1.1.1 sentence-transformers==5.1.0 -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 563.4/563.4 kB 25.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 139.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 78.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 99.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 71.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 111.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.9/19.9 MB 118.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 483.4/483.4 kB 43.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 26.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 86.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 458.9/458.9 kB 42.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.4

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [ ]:
#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

#Libraries for downloading and loading the llm
from huggingface_hub import hf_hub_download
from llama_cpp import Llama

## **Question Answering using LLM**

##### Downloading and Loading the model

In [ ]:
# Download LLM model from HuggingFace Hub

model_name_or_path ="TheBloke/Mistral-7B-Instruct-v0.2-GGUF"
model_basename = "mistral-7b-instruct-v0.2.Q6_K.gguf"

model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename = model_basename
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


mistral-7b-instruct-v0.2.Q6_K.gguf:   0%|          | 0.00/5.94G [00:00<?, ?B/s]

In [ ]:
# Load LLM model into memory using llama cpp, runtime is connected to GPU.
llm = Llama(
    model_path=model_path,
    n_ctx = 3500,
    n_gpu_layers = 38,
    n_batch=512
)


AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 


###**Observations:**



* For this analysis, I have used the "mistral-7b-instruct-v0.2.Q6_K.gguf" LLM
which is a 6 bit quantized model with 7 Billion paramaters.

* The max token chosen for this LLM is 3500, with 38 gpu layers and a batch size of 512.

##**Response**

In [ ]:
def response(query,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )

    return model_output['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
response(user_input)

'\n\nSepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following are general steps for managing sepsis in a critical care unit:\n\n1. Early recognition: Recognize the signs and symptoms of sepsis early and initiate treatment as soon as possible. Sepsis can present with various clinical features, including fever or hypothermia, tachycardia or bradycardia, altered mental status, respiratory distress, and lactic acidosis.\n2. ABCs'

####Output:

Sepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following are general steps for managing sepsis in a critical care unit:

1. Early recognition: Recognize the signs and symptoms of sepsis early and initiate treatment as soon as possible. Sepsis can present with various clinical features, including fever or hypothermia, tachycardia or bradycardia, altered mental status, respiratory distress, and lactic acidosis.
2. ABCs

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
user_input = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
response(user_input)

Llama.generate: prefix-match hit


'\n\nAppendicitis is a medical condition characterized by inflammation of the appendix, a small pouch-like structure that extends from the large intestine. The symptoms of appendicitis can vary from person to person, but some common signs include:\n\n1. Abdominal pain: The pain is typically located in the lower right side of the abdomen and may be constant or come and go. It may start as a mild discomfort that worsens over time.\n2. Loss of appetite: People with appendicitis often lose their appetite due to abdominal pain and nausea'

####Output:

Appendicitis is a medical condition characterized by inflammation of the appendix, a small pouch-like structure that extends from the large intestine. The symptoms of appendicitis can vary from person to person, but some common signs include:

1. Abdominal pain: The pain is typically located in the lower right side of the abdomen and may be constant or come and go. It may start as a mild discomfort that worsens over time.
2. Loss of appetite: People with appendicitis often lose their appetite due to abdominal pain and nausea

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
user_input = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
response(user_input)

Llama.generate: prefix-match hit


"\n\nSudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, and eyelashes.\n\nThe exact cause of alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include stress, genetics, viral infections, and certain medications."

####Output:
Sudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, and eyelashes.

The exact cause of alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include stress, genetics, viral infections, and certain medications.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
user_input = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
response(user_input)

Llama.generate: prefix-match hit


'\n\nA person who has sustained a physical injury to the brain tissue may require various treatments depending on the severity and location of the injury. Here are some common treatments that may be recommended:\n\n1. Emergency care: In case of a traumatic brain injury (TBI), it is essential to seek emergency medical attention as soon as possible. The primary goal of emergency care is to prevent further damage to the brain, stabilize vital signs, and manage any life-threatening conditions.\n2. Medications: Depending on the symptoms, healthcare professionals may prescribe medications to manage various conditions associated with a'

####Output:

A person who has sustained a physical injury to the brain tissue may require various treatments depending on the severity and location of the injury. Here are some common treatments that may be recommended:

1. Emergency care: In case of a traumatic brain injury (TBI), it is essential to seek emergency medical attention as soon as possible. The primary goal of emergency care is to prevent further damage to the brain, stabilize vital signs, and manage any life-threatening conditions.
2. Medications: Depending on the symptoms, healthcare professionals may prescribe medications to manage various conditions associated with a

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
user_input = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
response(user_input)

Llama.generate: prefix-match hit


"\n\nFirst and foremost, if you suspect that someone has fractured their leg while hiking, it's essential to ensure their safety and prevent further injury. Here are some necessary precautions:\n\n1. Keep the person calm and still: Encourage them to remain as still as possible to minimize pain and prevent worsening the injury.\n2. Assess the situation: Check for any signs of shock, such as pale skin, rapid heartbeat, or shallow breathing. If you notice these symptoms, seek medical help immediately.\n3. Immobilize the leg: Use a splint, sl"

####Output:
First and foremost, if you suspect that someone has fractured their leg while hiking, it's essential to ensure their safety and prevent further injury. Here are some necessary precautions:

1. Keep the person calm and still: Encourage them to remain as still as possible to minimize pain and prevent worsening the injury.
2. Assess the situation: Check for any signs of shock, such as pale skin, rapid heartbeat, or shallow breathing. If you notice these symptoms, seek medical help immediately.
3. Immobilize the leg: Use a splint, sl

### **Observations:**
* I have assigned a max token of 128, temperature as 0, top_k as 50 and top_p as 0.5 as the arameters for the LLM responses. We can expect a shorter answers with little to no creativity.

* From the LLM's answers we can see that this model did a fair job of providing answers that are contextually relevant. However, the answers are not concise and doesn't fully address all the questions. The reason could be the limited token size of 128 and temperature setting of 0, which could've reduced the creativity and summarization ability of the model.

## Question Answering using LLM with Prompt Engineering

Adding a system prompt to give the LLM mode instructions on how to retrieve and structure the answers.

In [ ]:
system_prompt = """You are a knowledgeable medical assistant. Your job is to read the report and provide clinically accurate and concise answers to user questions.
 If the report does not contain sufficient information to answer the question, just say you don't have enough information. Don't infer beyond the report or fabricate details."""

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
user_input = system_prompt + "\n" + "What is the protocol for managing sepsis in a critical care unit?"
response(user_input)

Llama.generate: prefix-match hit


"\nBased on the information provided in the report, the protocol for managing sepsis in a critical care unit typically includes the following steps:\n1. Early recognition and assessment of sepsis using validated scoring systems such as Sequential Organ Failure Assessment (SOFA) or Quick Sequential Organ Failure Assessment (qSOFA).\n2. Initiation of broad-spectrum antibiotics as soon as possible, based on the patient's clinical presentation and culture results if available.\n3. Fluid resuscitation to maintain adequate tissue perfusion and organ function, with careful monitoring of fluid balance and"

####Output:
Based on the information provided in the report, the protocol for managing sepsis in a critical care unit typically includes the following steps:
1. Early recognition and assessment of sepsis using validated scoring systems such as Sequential Organ Failure Assessment (SOFA) or Quick Sequential Organ Failure Assessment (qSOFA).
2. Initiation of broad-spectrum antibiotics as soon as possible, based on the patient's clinical presentation and culture results if available.
3. Fluid resuscitation to maintain adequate tissue perfusion and organ function, with careful monitoring of fluid balance and

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
user_input = system_prompt + "\n" + "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
response(user_input)

Llama.generate: prefix-match hit


'\nAccording to the report, the patient presented with the following symptoms: right lower quadrant abdominal pain, rebound tenderness, guarding, and decreased bowel sounds. These are common symptoms of appendicitis. Appendicitis is typically treated via surgical removal of the appendix, known as an appendectomy. Antibiotics alone may not be sufficient to cure appendicitis due to the risk of rupture and potential for peritonitis, a serious complication. Therefore, an appendectomy is the standard treatment for appendicitis.'

####Output:
According to the report, the patient presented with the following symptoms: right lower quadrant abdominal pain, rebound tenderness, guarding, and decreased bowel sounds. These are common symptoms of appendicitis. Appendicitis is typically treated via surgical removal of the appendix, known as an appendectomy. Antibiotics alone may not be sufficient to cure appendicitis due to the risk of rupture and potential for peritonitis, a serious complication. Therefore, an appendectomy is the standard treatment for appendicitis.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
user_input = system_prompt + "\n" + "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
response(user_input)

Llama.generate: prefix-match hit


'\nBased on the information provided in the report, there is no specific mention of the cause or treatment for sudden patchy hair loss. However, I can provide some general information on possible causes and treatments for this condition.\nPossible causes of sudden patchy hair loss include:\n1. Alopecia Areata: An autoimmune disorder that causes hair loss in round patches on the scalp, face, or other areas of the body.\n2. Trauma or stress: Physical or emotional trauma can cause sudden hair loss.\n3. Nutritional deficiencies: Lack of certain nutrients'

####Output:
Based on the information provided in the report, there is no specific mention of the cause or treatment for sudden patchy hair loss. However, I can provide some general information on possible causes and treatments for this condition.
Possible causes of sudden patchy hair loss include:
1. Alopecia Areata: An autoimmune disorder that causes hair loss in round patches on the scalp, face, or other areas of the body.
2. Trauma or stress: Physical or emotional trauma can cause sudden hair loss.
3. Nutritional deficiencies: Lack of certain nutrients

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
user_input = system_prompt + "\n" + "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
response(user_input)

Llama.generate: prefix-match hit


"\nBased on the information provided, I cannot determine the specific treatments recommended for a person with a brain injury as the report does not include that detail. However, common treatments for brain injuries may include medication to manage symptoms such as pain, swelling, or seizures; physical therapy and rehabilitation to help improve function and mobility; speech therapy to address communication difficulties; occupational therapy to help with daily living skills; and cognitive therapy to address memory and thinking problems. The specific treatment plan would depend on the severity and location of the injury, as well as the individual's overall health and response to treatment."

####Output:
Based on the information provided, I cannot determine the specific treatments recommended for a person with a brain injury as the report does not include that detail. However, common treatments for brain injuries may include medication to manage symptoms such as pain, swelling, or seizures; physical therapy and rehabilitation to help improve function and mobility; speech therapy to address communication difficulties; occupational therapy to help with daily living skills; and cognitive therapy to address memory and thinking problems. The specific treatment plan would depend on the severity and location of the injury, as well as the individual's overall health and response to treatment.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
user_input = system_prompt + "\n" + "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
response(user_input)

Llama.generate: prefix-match hit


'\nBased on the information provided, I cannot determine the specific type or location of the leg fracture. However, I can provide general guidelines for the care and treatment of a leg fracture.\n1. Immobilization: The leg should be immobilized using a cast, splint, or brace to prevent movement and promote healing. This will also help reduce pain and swelling.\n2. Pain management: Over-the-counter pain medications such as acetaminophen or ibuprofen can be used to manage pain. Prescription pain medication may be necessary in some cases.\n3. Elevation'

####Output:
Based on the information provided, I cannot determine the specific type or location of the leg fracture. However, I can provide general guidelines for the care and treatment of a leg fracture.
1. Immobilization: The leg should be immobilized using a cast, splint, or brace to prevent movement and promote healing. This will also help reduce pain and swelling.
2. Pain management: Over-the-counter pain medications such as acetaminophen or ibuprofen can be used to manage pain. Prescription pain medication may be necessary in some cases.
3. Elevation

###**Observations:**
* In this section I have added a system prompt to give additional instructions to define the content, tone and boundaries for the LLM's response while retaining the same parameters the was used in the previous LLM settings.

* We can observe that LLM's responses are relatively direct and relevant, focusing on key points without making up extra details compared to the previous set of responses. The answers are less creative as the LLM was instructed to follow the manual closely.

* Answer1: The LLM was able to answer to the context however, it was not able to provide all the key points due to limited token size and also not very good at summarizing the important points concisely.

* Answer2: The LLM did a good job of answering precisely to the all the questions and it also correctly suggested the surgical treatment for that condition. The answer appears grounded and relevant to the context.

* For answers 3, 4, and 5, the LLM experienced difficulty in matching the most accurate and relevant response to the questions. The answers provided were more generic in their discussion of context, but failed to retrieve responses that were specific to the given conditions.

## **Data Preparation for RAG**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### **Loading the Data**

In [ ]:
manual_pdf_path ="/content/drive/MyDrive/Colab Notebooks/medical_diagnosis_manual.pdf"

In [ ]:
pdf_loader = PyMuPDFLoader(manual_pdf_path)

In [ ]:
manual = pdf_loader.load()

### Data Overview

#### Checking the first 5 pages

In [ ]:
for i in range(5):
  print(f"Page number:{i+1}",end="\n")
  print(manual[i].page_content,end="\n")

Page number:1
sukanya.raj@gmail.com
AEFMT1S5GK
eant for personal use by sukanya.raj@gm
shing the contents in part or full is liable
Page number:2
sukanya.raj@gmail.com
AEFMT1S5GK
This file is meant for personal use by sukanya.raj@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.
Page number:3
Table of Contents
1
Front    ................................................................................................................................................................................................................
1
Cover    .......................................................................................................................................................................................................
2
Front Matter    ...........................................................................................................................................................................................
53
1 -

#### Checking the number of pages

In [ ]:
len(manual)

4114

### **Observations:**
* I have loaded the Merck manual PDF in order to train the RAG model with medical data in order to generate answers relevant to healthcare related questions. A total of 4114 pages of data were loaded to train the model.

### Data Chunking

In [ ]:
text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name="cl100k_base",
    chunk_size = 512,
    chunk_overlap = 64
)

In [ ]:
#Split the text into chunks
document_chunks = pdf_loader.load_and_split(text_splitter)

In [ ]:
len(document_chunks)

8781

In [ ]:
document_chunks[0].page_content

'sukanya.raj@gmail.com\nAEFMT1S5GK\neant for personal use by sukanya.raj@gm\nshing the contents in part or full is liable'

In [ ]:
document_chunks[1].page_content

'sukanya.raj@gmail.com\nAEFMT1S5GK\nThis file is meant for personal use by sukanya.raj@gmail.com only.\nSharing or publishing the contents in part or full is liable for legal action.'

In [ ]:
document_chunks[2].page_content

'Table of Contents\n1\nFront    ................................................................................................................................................................................................................\n1\nCover    .......................................................................................................................................................................................................\n2\nFront Matter    ...........................................................................................................................................................................................\n53\n1 - Nutritional Disorders    ...............................................................................................................................................................\n53\nChapter 1. Nutrition: General Considerations    ...........................................................................................

### Embedding

In [ ]:
embedding_model = SentenceTransformerEmbeddings(model_name="thenlper/gte-large")

embeddings1 = embedding_model.embed_query(document_chunks[0].page_content)
embeddings2 = embedding_model.embed_query(document_chunks[1].page_content)

embeddings1 , embeddings2


/tmp/ipython-input-100213165.py:1: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_model = SentenceTransformerEmbeddings(model_name="thenlper/gte-large")


modules.json:   0%|          | 0.00/385 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/619 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/670M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/342 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

([-0.03296686336398125,
  -0.007466697599738836,
  -0.024299051612615585,
  0.0011634787078946829,
  -0.007388183381408453,
  -0.04247644171118736,
  0.004433322697877884,
  0.03785345330834389,
  0.01506460178643465,
  0.008257897570729256,
  0.02827160432934761,
  -0.011313369497656822,
  0.011375908739864826,
  -0.029989467933773994,
  -0.02536381408572197,
  0.005625016521662474,
  -0.006268042139708996,
  -0.04113604128360748,
  0.0018899457063525915,
  0.0023509629536420107,
  0.019001733511686325,
  0.022116363048553467,
  -0.09962479025125504,
  -0.03608693927526474,
  -0.0011627892963588238,
  0.016387896612286568,
  0.018338356167078018,
  -0.005641631316393614,
  0.0767456516623497,
  0.038917724043130875,
  -0.0312258992344141,
  -0.009170908480882645,
  0.012346498668193817,
  -0.04075547307729721,
  -0.012639629654586315,
  0.006407756824046373,
  0.03461547568440437,
  -0.020070653408765793,
  -0.020967792719602585,
  -0.045389555394649506,
  -0.003942752256989479,
  -0.

In [ ]:
print("Dimensions of the embedding vector:", len(embeddings1))
len(embeddings1) == len(embeddings2)


Dimensions of the embedding vector: 1024


True

###**Observations:**
* I have used a recurssivecharactertext splitter to split the documents into data chunks of size 512 and an overlap of 64 characters. From the retrieved chunks we can see some overlapped characters.

* I've used the **"Thenlper/gte-large"** model as the sentencetransformer model for emddebing the data chunks. Each embedding vector is of dimension 1024.

### Vector Database

In [ ]:
out_dir = 'medical_db'

if not os.path.exists(out_dir):
  os.makedirs(out_dir)

In [ ]:
vectorstore = Chroma.from_documents(
    document_chunks,
    embedding_model,
    persist_directory = out_dir
)

In [ ]:
vectorstore = Chroma(persist_directory=out_dir,embedding_function=embedding_model)

/tmp/ipython-input-2756559696.py:1: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-chroma package and should be used instead. To use it run `pip install -U :class:`~langchain-chroma` and import as `from :class:`~langchain_chroma import Chroma``.
  vectorstore = Chroma(persist_directory=out_dir,embedding_function=embedding_model)


In [ ]:
vectorstore.embeddings

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 512, 'do_lower_case': False, 'architecture': 'BertModel'})
  (1): Pooling({'word_embedding_dimension': 1024, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
), model_name='thenlper/gte-large', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False, show_progress=False)

In [ ]:
vectorstore.similarity_search("abdominal pain",k=3)

[Document(metadata={'creationdate': '2012-06-15T05:44:40+00:00', 'trapped': '', 'moddate': '2026-01-02T01:09:23+00:00', 'keywords': '', 'total_pages': 4114, 'file_path': '/content/drive/MyDrive/Colab Notebooks/medical_diagnosis_manual.pdf', 'author': '', 'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'creator': 'Atop CHM to PDF Converter', 'creationDate': 'D:20120615054440Z', 'page': 167, 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'format': 'PDF 1.7', 'subject': '', 'source': '/content/drive/MyDrive/Colab Notebooks/medical_diagnosis_manual.pdf', 'modDate': 'D:20260102010923Z'}, page_content='Chapter 11. Acute Abdomen and Surgical Gastroenterology\nIntroduction\nAcute abdomen refers to abdominal symptoms and signs of such severity or concern that disorders\nrequiring surgery should be considered. The primary symptom is acute abdominal pain. Chronic\nabdominal pain is discussed in Ch. 7.\nAcute Abdominal Pain\nAbdominal pain is common and often inconse

1. page_content='Chapter 11. Acute Abdomen and Surgical Gastroenterology\nIntroduction\nAcute abdomen refers to abdominal symptoms and signs of such severity or concern that disorders\nrequiring surgery should be considered. The primary symptom is acute abdominal pain. Chronic\nabdominal pain is discussed in Ch. 7.\nAcute Abdominal Pain\nAbdominal pain is common and often inconsequential. Acute and severe abdominal pain, however, is\nalmost always a symptom of intra-abdominal disease. It may be the sole indicator of the need for surgery\nand must be attended to swiftly: Gangrene and perforation of the gut can occur < 6 h from onset of\nsymptoms in certain conditions (eg, interruption of the intestinal blood supply caused by a strangulating\nobstruction or an arterial embolus). Abdominal pain is of particular concern in patients who are very young\nor very old and those who have HIV infection or are taking immunosuppressants.\nTextbook descriptions of abdominal pain have limitations because people react to pain differently. Some,\nparticularly elderly people, are stoic, whereas others exaggerate their symptoms. Infants, young children,\nand some elderly people may have difficulty localizing the pain.\nPathophysiology\nVisceral pain comes from the abdominal viscera, which are innervated by autonomic nerve fibers and\nrespond mainly to the sensations of distention and muscular contraction—not to cutting, tearing, or local\nirritation. Visceral pain is typically vague, dull, and nauseating. It is poorly localized and tends to be\nreferred to areas corresponding to the embryonic origin of the affected structure. Foregut structures\n(stomach, duodenum, liver, and pancreas) cause upper abdominal pain. Midgut structures (small bowel,\nproximal colon, and appendix) cause periumbilical pain. Hindgut structures (distal colon and GU tract)\ncause lower abdominal pain.\nSomatic pain comes from the parietal peritoneum, which is innervated by somatic nerves, which respond\nto irritation from infectious, chemical, or other inflammatory processes. Somatic pain is sharp and well\nlocalized.\nReferred pain is pain perceived distant from its source and results from convergence of nerve fibers at\nthe spinal cord. Common examples of referred pain are scapular pain due to biliary colic, groin pain due'),

2. page_content='Fig. 11-1); some are trivial but some are immediately life threatening, requiring rapid diagnosis and\nsurgery. These include ruptured abdominal aortic aneurysm (AAA), perforated viscus, mesenteric\nischemia, and ruptured ectopic pregnancy. Others (eg, intestinal obstruction, appendicitis, severe acute\npancreatitis) are also serious and nearly as urgent. Several extra-abdominal disorders also cause\nabdominal pain (see\nTable 11-1).\nAbdominal pain in neonates, infants, and young children has numerous causes not encountered in adults,\nincluding meconium peritonitis, pyloric stenosis, esophageal webs, volvulus of a gut with a common\nmesentery, imperforate anus, intussusception, and intestinal obstruction caused by atresia.\nEvaluation\nEvaluation of mild and severe pain follows the same process, although with severe abdominal pain,\ntherapy sometimes proceeds simultaneously and involves early consultation with a surgeon. History and\nphysical examination usually exclude all but a few possible causes, with final diagnosis confirmed by\njudicious use of laboratory and imaging tests. Life-threatening causes should always be ruled out before\nfocusing on less serious diagnoses. In seriously ill patients with severe abdominal pain, the most\nimportant diagnostic measure may be expeditious surgical exploration. In mildly ill patients, watchful\nwaiting may be best.\nHistory: A thorough history usually suggests the diagnosis (see\nTable 11-2). Of particular importance are pain location (see Fig. 11-1) and characteristics, history of\nsimilar symptoms, and associated symptoms. Concomitant symptoms such as gastroesophageal reflux,\nnausea, vomiting, diarrhea, constipation, jaundice, melena, hematuria, hematemesis, weight loss, and\nmucus or blood in the stool help direct subsequent evaluation. A drug history should include details\nconcerning prescription and illicit drug use as well as alcohol. Many drugs cause GI upset. Prednisone or\nimmunosuppressants may inhibit the inflammatory response to perforation or peritonitis and result in less\npain and leukocytosis than might otherwise be expected. Anticoagulants can increase the chances of\nbleeding and hematoma formation. Alcohol predisposes to pancreatitis.\nKnown medical conditions and previous abdominal surgeries are important to ascertain. Women should'),


3. page_content="with renal colic may be unable to lie still and may pace, writhe, or constantly shift position. The abdomen\nmay be somewhat tender on the affected side as palpation increases pressure in the already-distended\nureter, but peritoneal signs (guarding, rebound, rigidity) are lacking. For some patients, the first symptom\nis hematuria or either gravel or a calculus in the urine. Other patients may have symptoms of a UTI, such\nas fever, dysuria, or cloudy or foul-smelling urine.\nDiagnosis\n• Clinical differential diagnosis\n• Urinalysis\n• Imaging\n• Determination of calculus composition\nThe symptoms and signs suggest the diagnosis. With peritonitis (eg, due to appendicitis, ectopic\npregnancy, or pelvic inflammatory disease), pain is usually constant, and patients lie still because\nmovement worsens pain. Patients often have rebound tenderness or rigidity. Cholecystitis may cause\ncolicky pain, usually in the epigastrium or right upper quadrant, often with Murphy's sign. Bowel\nobstruction may cause colicky abdominal pain and vomiting, but the pain is usually bilateral and not\nlocated primarily in the flank or along the ureter. Pancreatitis may cause upper abdominal pain and\nvomiting, but the pain is usually constant, may be bilateral, and is usually not along the flank or ureter.\nWith most of these disorders, urinary symptoms are uncommon and other symptoms may suggest which\norgan system is actually involved (eg, vaginal discharge or bleeding in pelvic disorders among females).\nDissecting aortic aneurysm must be considered, particularly in the elderly, because, if a renal artery is\naffected, it can cause hematuria, pain that radiates along a ureteral distribution, or both. Other\nThe Merck Manual of Diagnosis & Therapy, 19th Edition\nChapter 232. Urinary Calculi

###**Observations:**

A similarity check with the embeddings in the vector database for abdominal pain has given us three responses (k=3). We can see that responses have some generic context that matches the words abdomen and pain.

### **Retriever**

In [ ]:
retriever = vectorstore.as_retriever(
    search_type="similarity",
    search_kwargs={"k":2}
)

In [ ]:
rel_docs = retriever.get_relevant_documents("What are the treatments for Irritable Bowel Syndrome?")
rel_docs

/tmp/ipython-input-4260487767.py:1: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 1.0. Use :meth:`~invoke` instead.
  rel_docs = retriever.get_relevant_documents("What are the treatments for Irritable Bowel Syndrome?")


[Document(metadata={'subject': '', 'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'creator': 'Atop CHM to PDF Converter', 'source': '/content/drive/MyDrive/Colab Notebooks/medical_diagnosis_manual.pdf', 'format': 'PDF 1.7', 'trapped': '', 'file_path': '/content/drive/MyDrive/Colab Notebooks/medical_diagnosis_manual.pdf', 'keywords': '', 'modDate': 'D:20260102010923Z', 'moddate': '2026-01-02T01:09:23+00:00', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'page': 236, 'creationDate': 'D:20120615054440Z', 'total_pages': 4114, 'author': '', 'creationdate': '2012-06-15T05:44:40+00:00'}, page_content="Therapy is directed at specific symptoms. An effective therapeutic relationship is essential for effectively\nmanaging IBS. Patients should be invited to express not only their symptoms but also their understanding\nof their symptoms and the reasons prompting a visit to the health care practitioner (eg, fear of serious\ndisease). Patients should be educated about

1. page_content="Therapy is directed at specific symptoms. An effective therapeutic relationship is essential for effectively\nmanaging IBS. Patients should be invited to express not only their symptoms but also their understanding\nof their symptoms and the reasons prompting a visit to the health care practitioner (eg, fear of serious\ndisease). Patients should be educated about the disorder (eg, normal bowel physiology and the bowel's\nhypersensitivity to stress and food) and reassured, after appropriate tests, about the absence of a serious\nor life-threatening disease. Appropriate therapeutic goals (eg, expectations regarding the normal course\nor variability in symptoms, adverse effects of drugs, the appropriate and available working relationship\nbetween the physician and the patient) should be established. Finally, patients can benefit by being\nactively involved in the management of their condition. When successful, this can enhance the patient's\nmotivation to adhere to treatment, foster a more positive physician-patient relationship, and mobilize the\ncoping resources of even the most chronically passive patients. Psychologic stress, anxiety, or mood\ndisorders should be identified, evaluated, and treated. Regular physical activity helps relieve stress and\nassists in bowel function, particularly in patients with constipation.\nDiet: In general, a normal diet should be followed. Meals should not be overly large, and eating should be\nslow and paced. Patients with abdominal distention and increased flatulence may benefit from reducing or\neliminating beans, cabbage, and other foods containing fermentable carbohydrates. Reduced intake of\nsweeteners (eg, sorbitol, mannitol, fructose), which are constituents of natural and processed foods (eg,\napple and grape juice, bananas, nuts, and raisins), may alleviate flatulence, bloating, and diarrhea.\nPatients with evidence of lactose intolerance should reduce their intake of milk and dairy products. A low-\nfat diet may reduce postprandial abdominal symptoms.\nDietary fiber supplements may soften stool and improve the ease of evacuation. A bland bulk-producing\nagent may be used (eg, raw bran, starting with 15 mL [1 tbsp] with each meal, supplemented with\nThe Merck Manual of Diagnosis & Therapy, 19th Edition\nChapter 18. Irritable Bowel Syndrome

2.  page_content='increasing as necessary and tolerated up to about 100 to 150 mg once/day. SSRIs are also useful,\nparticularly for patients with anxiety or an affective disorder, but may exacerbate diarrhea. 5HT3\nantagonists (eg, alosetron) may benefit female patients with severe diarrhea refractory to other drugs.\nBecause alosetron is associated with ischemic colitis, its use is restricted.\nPreliminary data suggest that certain probiotics (eg, Bifidobacterium infantis) alleviate IBS symptoms,\nparticularly bloating. The beneficial effects of probiotics are not generic to the entire species but specific\nto certain strains. Certain aromatic oils (carminatives) can relax smooth muscle and relieve pain caused\nby cramps in some patients. Peppermint oil is the most commonly used agent in this class.\nPsychologic therapies: Cognitive-behavioral therapy, standard psychotherapy, and hypnotherapy may\nhelp some IBS patients.\nThe Merck Manual of Diagnosis & Therapy, 19th Edition\nChapter 18. Irritable Bowel Syndrome

###**Observations:**
* The two responses from the retriever matches the context of IBS but the answers appear to be generic and not necessarily talk about the actual treatments for Irritable Bowel Syndrome.

* This can be addressed by giving more details for the model to answer specific requirements using system prompts.

In [ ]:
model_output = llm(
      prompt=rel_docs[0].page_content,
      max_tokens=128,
      temperature=0,
      top_p=0.95,
      top_k=50
    )

Llama.generate: prefix-match hit


In [ ]:
model_output['choices'][0]['text']

' Any other use without permission of the author is strictly prohibited.\nwater if needed). Psyllium (Metamucil) may be used in a dose of 1 to 2 tbsp daily, divided into 3 doses.\nPatients with diarrhea should avoid caffeine and alcohol, which can worsen symptoms. They also should\navoid foods that are high in fat or fiber content (eg, fried foods, beans, corn, and whole grains).\nProbiotics: Probiotics may be helpful for patients with IBS who have diarrhea'

 Any other use without permission of the author is strictly prohibited.
water if needed). Psyllium (Metamucil) may be used in a dose of 1 to 2 tbsp daily, divided into 3 doses.
Patients with diarrhea should avoid caffeine and alcohol, which can worsen symptoms. They also should
avoid foods that are high in fat or fiber content (eg, fried foods, beans, corn, and whole grains).
Probiotics: Probiotics may be helpful for patients with IBS who have diarrhea

###**Observations:**
Here, I've included fine-tuning parameters like max_tones, temperature, top_k, and top_p to keep summaries brief and limit irrelevant information by minimizing creativity.

### System and User Prompt Template

In [ ]:
qna_system_message = """You are an assistant whose work is to review the report and provide the appropriate answers from the context.
User input will have the context required by you to answer user questions.
This context will begin with the token: ###Context.
The context contains references to specific portions of a document relevant to the user query.

User questions will begin with the token: ###Question.

Please answer only using the context provided in the input. Do not mention anything about the context in your final answer.

If the answer is not found in the context, respond "I don't know"."""

In [ ]:
qna_user_message_template= """###Context
Here are some documents that are relevant to the question mentioned below.
{context}

###Question
{question}"""

### Response Function

In [ ]:
def generate_rag_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

### **Observations:**

* I've add a system prompt that provides specific instructions to the model on how to tone, structure, respond to the user questions as a virtual medical assistant.

* I've also added a user template prompt to include the context and the user question.

## **Question Answering using RAG**

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
generate_rag_response(user_input,top_k=20)

Llama.generate: prefix-match hit


'Based on the context provided, the protocol for managing sepsis in a critical care unit includes:\n1. Treatment in an ICU with frequent monitoring of systemic pressure, CVP or PAOP, pulse oximetry, ABGs, blood glucose, lactate and electrolyte levels, renal function, urine output, and possibly sublingual PCO2.\n2. Fluid resuscitation with 0.9% saline until CVP reaches 8 mm Hg (10 cm H2O) or PAOP reaches 12 to 15'

####Output:
Based on the context provided, the protocol for managing sepsis in a critical care unit includes:
1. Treatment in an ICU with frequent monitoring of systemic pressure, CVP or PAOP, pulse oximetry, ABGs, blood glucose, lactate and electrolyte levels, renal function, urine output, and possibly sublingual PCO2.
2. Fluid resuscitation with 0.9% saline until CVP reaches 8 mm Hg (10 cm H2O) or PAOP reaches 12 to 15

###**Observations:**
* The repsonse is contextually relevant but incomplete. We can either reduce the top K value to get more concise and precise reponse. We can also increase the token size to get the full response.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
user_input = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
generate_rag_response(user_input,top_k=20)

Llama.generate: prefix-match hit


'The context mentions that appendicitis is acute inflammation of the vermiform appendix, typically resulting in abdominal pain, anorexia, and abdominal tenderness. The treatment for appendicitis is surgical removal. There is no mention of any medication being effective in curing appendicitis.'

####Output:

The context mentions that appendicitis is acute inflammation of the vermiform appendix, typically resulting in abdominal pain, anorexia, and abdominal tenderness. The treatment for appendicitis is surgical removal. There is no mention of any medication being effective in curing appendicitis.

###**Observations:**
Answer 2 is very precise and to the point. It appears to address all the questions in the context.
We can fine tune the temparature value to check if we can get some more additional details about this condition.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
user_input = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
generate_rag_response(user_input,top_k=20)

Llama.generate: prefix-match hit


'Based on the context provided, alopecia areata is a condition characterized by sudden patchy hair loss. The most common treatment options for this condition include topical corticosteroids in combination with a long-acting oral tetracycline, or topical minoxidil. Other possible treatments may include intralesional or systemic corticosteroids, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). The cause of alopecia areata is believed to'

####Output:
Based on the context provided, alopecia areata is a condition characterized by sudden patchy hair loss. The most common treatment options for this condition include topical corticosteroids in combination with a long-acting oral tetracycline, or topical minoxidil. Other possible treatments may include intralesional or systemic corticosteroids, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). The cause of alopecia areata is believed to

##**Observations:**
Again in answer 3 we can observe that the reponse is incomplete due to limited token size. We need to fine tune parameter to get a better and complete responses.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
user_input = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
generate_rag_response(user_input,top_k=20)

Llama.generate: prefix-match hit


"Based on the context provided, rehabilitation is recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function. The rehabilitation process involves a team approach that combines physical, occupational, and speech therapy, skill-building activities, and counseling to meet the patient's social and emotional needs. For patients whose coma exceeds 24 hours and have major persistent neurologic sequelae, a prolonged period of rehabilitation, particularly in cognitive and emotional areas, is often required. Rehabilitation services should be planned early."

####Output:

Based on the context provided, rehabilitation is recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function. The rehabilitation process involves a team approach that combines physical, occupational, and speech therapy, skill-building activities, and counseling to meet the patient's social and emotional needs. For patients whose coma exceeds 24 hours and have major persistent neurologic sequelae, a prolonged period of rehabilitation, particularly in cognitive and emotional areas, is often required. Rehabilitation services should be planned early.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
user_input = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
generate_rag_response(user_input,top_k=20)

Llama.generate: prefix-match hit


'The context mentions that fractures require sterile wound dressings, tetanus prophylaxis, and broad-spectrum antibiotics. The person should have the affected area immobilized with splints or bulky dressings, and wounds near joints should be elevated above heart level for the first 48 hours after suturing. Wounds are kept clean and dry, and nonadherent and impermeable to bacteria dressings are applied. Antibiotic ointment is used daily until the wound closure device is removed. If an odor emanates from within'

####Output:
The context mentions that fractures require sterile wound dressings, tetanus prophylaxis, and broad-spectrum antibiotics. The person should have the affected area immobilized with splints or bulky dressings, and wounds near joints should be elevated above heart level for the first 48 hours after suturing. Wounds are kept clean and dry, and nonadherent and impermeable to bacteria dressings are applied. Antibiotic ointment is used daily until the wound closure device is removed. If an odor emanates from within

### **Observations:**

* We can observe from the above RAG responses that by providing more details and context through system and user prompts, the model generates more accurate and concise answers without including irrelevant information. Because the model is not fine-tuned, answers 1,3 and 5 are incomplete due to limited token size. This could be improved by increasing max_tokens or adjusting the temperature for better summarization.

### **Fine-tuning**

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
user_input = "What is the protocol for managing sepsis in a critical care unit?"
generate_rag_response(user_input,temperature=0,max_tokens=650, top_p=.95)

Llama.generate: prefix-match hit


"Based on the context provided, the protocol for managing sepsis in a critical care unit includes:\n1. Treatment in an ICU with frequent monitoring of systemic pressure, CVP or PAOP, pulse oximetry, ABGs, blood glucose, lactate and electrolyte levels, renal function, urine output, and possibly sublingual PCO2.\n2. Fluid resuscitation with 0.9% saline until CVP reaches 8 mm Hg (10 cm H2O) or PAOP reaches 12 to 15 mm Hg. Oliguria with hypotension is not a contraindication to vigorous fluid resuscitation.\n3. Administration of oxygen by mask or nasal prongs, and possible tracheal intubation and mechanical ventilation for respiratory failure.\n4. Use of dopamine to increase mean BP to at least 60 mm Hg if the patient remains hypotensive after CVP or PAOP has been raised to target levels. If dopamine dose exceeds 20 μg/kg/min, another vasopressor like norepinephrine may be added.\n5. Point-of-care testing for certain blood tests at the patient's bedside or unit.\n6. Administration of parent

####Output:
Based on the context provided, the protocol for managing sepsis in a critical care unit includes:
1. Treatment in an ICU with frequent monitoring of systemic pressure, CVP or PAOP, pulse oximetry, ABGs, blood glucose, lactate and electrolyte levels, renal function, urine output, and possibly sublingual PCO2.
2. Fluid resuscitation with 0.9% saline until CVP reaches 8 mm Hg (10 cm H2O) or PAOP reaches 12 to 15 mm Hg. Oliguria with hypotension is not a contraindication to vigorous fluid resuscitation.
3. Administration of oxygen by mask or nasal prongs, and possible tracheal intubation and mechanical ventilation for respiratory failure.
4. Use of dopamine to increase mean BP to at least 60 mm Hg if the patient remains hypotensive after CVP or PAOP has been raised to target levels. If dopamine dose exceeds 20 μg/kg/min, another vasopressor like norepinephrine may be added.
5. Point-of-care testing for certain blood tests at the patient's bedside or unit.
6. Administration of parenteral antibiotics after specimens have been taken for Gram stain and culture, with empiric therapy started immediately after suspecting sepsis. Antibiotic selection requires an educated guess based on the suspected source, clinical setting, knowledge or suspicion of causative organisms and sensitivity patterns common to that specific inpatient unit, and previous culture results.
7. Drainage of abscesses and excision of necrotic tissue, such as infarcted bowel, gangrenous gall-bladder, or abscessed uterus.
8. Normalization of blood glucose levels with a continuous IV insulin infusion titrated to maintain glucose between 80 to 110 mg/dL (4.4 to 6.1 mmol/L).

###**Observations**:
Answer 1: I fine-tuned the model using the parameters temperature = 0, max_tokens = 650, and top_p = 0.95. My objective was to generate responses that closely align with the source material and provide comprehensive coverage of all key aspects of the questions, as previous RAG responses were incomplete. The selected temperature and top_p values help ensure the output remains accurate and relevant to the established source of truth.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
user_input = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
generate_rag_response(user_input,temperature=0.5, top_k=10)

Llama.generate: prefix-match hit


'The common symptoms for appendicitis include abdominal pain, anorexia, and abdominal tenderness. The document suggests that treatment is surgical removal of the appendix. No mention was made about curing appendicitis via medicine in the context.'

####Output:
The common symptoms for appendicitis include abdominal pain, anorexia, and abdominal tenderness. The document suggests that treatment is surgical removal of the appendix. No mention was made about curing appendicitis via medicine in the context.

###**Observations:**
Answer 2: Setting temperature to 0.5 and top_k to 10 allowed the model to generate concise, complete responses by limiting token selection and enhancing summarization ability.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
user_input = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
generate_rag_response(user_input,temperature=0.1, max_tokens=250, top_k=20)


Llama.generate: prefix-match hit


'Based on the context provided, the condition being described is alopecia areata. The effective treatments for this condition include topical corticosteroids in combination with a long-acting oral tetracycline for scarring forms, or topical minoxidil for non-scarring forms. Other treatment options may include intralesional or systemic corticosteroids, topical immunotherapy, or antimalarials for lichen planopilaris and chronic cutaneous lupus lesions. It is important to note that the underlying cause of alopecia areata is an autoimmune disorder affecting genetically susceptible individuals exposed to unclear environmental triggers. Therefore, identifying and addressing any potential triggers may also be necessary for effective treatment.'

####Output:
Based on the context provided, the condition being described is alopecia areata. The effective treatments for this condition include topical corticosteroids in combination with a long-acting oral tetracycline for scarring forms, or topical minoxidil for non-scarring forms. Other treatment options may include intralesional or systemic corticosteroids, topical immunotherapy, or antimalarials for lichen planopilaris and chronic cutaneous lupus lesions. It is important to note that the underlying cause of alopecia areata is an autoimmune disorder affecting genetically susceptible individuals exposed to unclear environmental triggers. Therefore, identifying and addressing any potential triggers may also be necessary for effective treatment.

###**Observations:**

Answer 3: Using temperature=0.1, max_tokens=250, and top_k=20 produced a response that accurately covers all key points.


### Query 4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
user_input = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
generate_rag_response(user_input,max_tokens=250, top_k=20)

Llama.generate: prefix-match hit


"Based on the context provided, rehabilitation is recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function. The rehabilitation process involves a team approach that combines physical, occupational, and speech therapy, skill-building activities, and counseling to meet the patient's social and emotional needs. For patients whose coma exceeds 24 hours and have major persistent neurologic sequelae, a prolonged period of rehabilitation, particularly in cognitive and emotional areas, is often required. Rehabilitation services should be planned early."

####Output:

Based on the context provided, rehabilitation is recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function. The rehabilitation process involves a team approach that combines physical, occupational, and speech therapy, skill-building activities, and counseling to meet the patient's social and emotional needs. For patients whose coma exceeds 24 hours and have major persistent neurologic sequelae, a prolonged period of rehabilitation, particularly in cognitive and emotional areas, is often required. Rehabilitation services should be planned early.

###**Observations:**
By setting max_tokens to 250 and top_k to 20 for Answer 4, the model is able to produce detailed and accurate responses that address treatments as well as both short-term and long-term effects.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
user_input = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
generate_rag_response(user_input,temperature=0.1,max_tokens=350, top_p=0.8)


Llama.generate: prefix-match hit


'The context mentions that fractures require sterile wound dressings, tetanus prophylaxis, and broad-spectrum antibiotics. The person should also have the affected area immobilized with splints or bulky dressings, and the wound elevated above heart level for the first 48 hours after suturing. Antibiotics are given as early as possible, and the patient is advised to maintain meticulous wound care by keeping the wound clean and dry and applying nonadherent and impermeable dressings. The person should also be instructed to inspect minor, clean lacerations but seek medical care for higher risk wounds or wounds in unreliable patients. For fractures that require immobilization for several days or less, a splint can be used to allow the patient to apply ice and move more without contributing to compartment syndrome. Prolonged immobilization (more than 3 to 4 weeks) of a joint can cause stiffness, contractures, and muscle atrophy. Some rapidly healing injuries are best treated with early mobil

####Output:

The context mentions that fractures require sterile wound dressings, tetanus prophylaxis, and broad-spectrum antibiotics. The person should also have the affected area immobilized with splints or bulky dressings, and the wound elevated above heart level for the first 48 hours after suturing. Antibiotics are given as early as possible, and the patient is advised to maintain meticulous wound care by keeping the wound clean and dry and applying nonadherent and impermeable dressings. The person should also be instructed to inspect minor, clean lacerations but seek medical care for higher risk wounds or wounds in unreliable patients. For fractures that require immobilization for several days or less, a splint can be used to allow the patient to apply ice and move more without contributing to compartment syndrome. Prolonged immobilization (more than 3 to 4 weeks) of a joint can cause stiffness, contractures, and muscle atrophy. Some rapidly healing injuries are best treated with early mobilization to minimize contractures and muscle atrophy. The context also mentions that fractures are cracks in bones that may result from a single application of significant force or from a bone weakened by a disorder. Symptoms include pain, swelling, ecchymosis, crepitation, deformity, and abnormal motion. Occasional complications include fat embolism, arterial injury, compartment syndrome, nerve injuries, and infection. Diagnosis is by clinical criteria and usually plain x-rays. Treatment involves analgesics, immobilization

### **Observations:**

* Answer 5: Setting temperature to 0.1, max_tokens to 350, and top_p to 0.8 made the model produce more complete and relevant responses.

* Overall we can observe that fine-tuned RAG model produces more deterministic, focused, and precise responses than those without fine-tuning.

## Output Evaluation

Let us now use the LLM-as-a-judge method to check the quality of the RAG system on two parameters - retrieval and generation. We illustrate this evaluation based on the answeres generated to the question from the previous section.

- We are using the same Mistral model for evaluation, so basically here the llm is rating itself on how well he has performed in the task.

In [ ]:
groundedness_rater_system_message  = """
You are tasked with rating AI generated answers to questions posed by users.
You will be presented a question, context used by the AI system to generate the answer and an AI generated answer to the question.
In the input, the question will begin with ###Question, the context will begin with ###Context while the AI generated answer will begin with ###Answer.

Evaluation criteria:
The task is to judge the extent to which the metric is followed by the answer.
1 - The metric is not followed at all
2 - The metric is followed only to a limited extent
3 - The metric is followed to a good extent
4 - The metric is followed mostly
5 - The metric is followed completely

Metric:
The answer should be derived only from the information presented in the context

Instructions:
1. First write down the steps that are needed to evaluate the answer as per the metric.
2. Give a step-by-step explanation if the answer adheres to the metric considering the question and context as the input.
3. Next, evaluate the extent to which the metric is followed.
4. Use the previous information to rate the answer using the evaluaton criteria and assign a score.
"""

In [ ]:
relevance_rater_system_message = """
You are tasked with rating AI generated answers to questions posed by users.
You will be presented a question, context used by the AI system to generate the answer and an AI generated answer to the question.
In the input, the question will begin with ###Question, the context will begin with ###Context while the AI generated answer will begin with ###Answer.

Evaluation criteria:
The task is to judge the extent to which the metric is followed by the answer.
1 - The metric is not followed at all
2 - The metric is followed only to a limited extent
3 - The metric is followed to a good extent
4 - The metric is followed mostly
5 - The metric is followed completely

Metric:
Relevance measures how well the answer addresses the main aspects of the question, based on the context.
Consider whether all and only the important aspects are contained in the answer when evaluating relevance.

Instructions:
1. First write down the steps that are needed to evaluate the context as per the metric.
2. Give a step-by-step explanation if the context adheres to the metric considering the question as the input.
3. Next, evaluate the extent to which the metric is followed.
4. Use the previous information to rate the context using the evaluaton criteria and assign a score.
"""

In [ ]:
user_message_template =  """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""

In [ ]:
def generate_ground_relevance_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=3)
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    answer =  response["choices"][0]["text"]

    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    # Combine user_prompt and system_message to create the prompt
    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    response_2 = llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    return response_1['choices'][0]['text'],response_2['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
ground,rel = generate_ground_relevance_response(user_input="What is the protocol for managing sepsis in a critical care unit?",max_tokens=1024)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the information in the context related to managing sepsis in a critical care unit.
2. Compare each point in the AI generated answer with the corresponding information in the context.
3. Determine if the AI generated answer is derived only from the information presented in the context.

Explanation:
The AI generated answer includes several points related to managing sepsis in a critical care unit, such as monitoring vital signs, quantifying fluid intake and output, measuring daily weight, recording blood pressure, routine daily blood tests, parenteral antibiotics, antibiotic selection, draining abscesses, normalization of blood glucose, and titrating insulin infusion.

The context provides information on the importance of critical care medicine in treating seriously ill patients, supportive care for ICU patients, patient monitoring and testing, blood tests, and managing sepsis specifically. The context mentions several aspects related to managi

###**Observations:**
The evaluation for answer 1 gave a rating of 5 for groundedness and rating of 4 for relevance. This indicates that the answer is taken from the verifiable source and that there are no hallucinations. However, a rating of 4 for relevance could indicate that some key points might be missing.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
ground,rel = generate_ground_relevance_response(user_input="What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?",max_tokens=350) #Complete the code to pass the query #2 along with parameters if needed

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the question parts: symptoms and treatment for appendicitis, and whether it can be cured via medicine.
2. Find relevant information in the context regarding symptoms and treatment of appendicitis.
3. Check if the answer is derived only from the information presented in the context.

Explanation:
The AI generated answer correctly identifies the common symptoms for appendicitis as abdominal pain, anorexia, and abdominal tenderness. It also states that the treatment for appendicitis is not curable via medicine alone, and surgical removal of the appendix (appendectomy) is required.

The answer adheres to the metric since it only includes information derived from the context. The symptoms mentioned in the question are explicitly stated in the context, as well as the treatment requirement for surgery. Therefore, the answer accurately reflects the information provided in the context.

Rating:
Based on the evaluation criteria, I would rate this answer

###**Observations:**
For answer 2, again the groundedness rating is 5 which indicates the answer is true to the source. The relevance rating is 4 , however the evaulation mentions that it does not contain irrelevant information. As there is no clear justification,the score could probably mean that the answer doesn't cover all the aspects of the question.

###**Observations:**


### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
ground,rel = generate_ground_relevance_response(user_input="What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?",max_tokens=350) #Complete the code to pass the query #3 along with parameters if needed

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the main topic of the question, which is about effective treatments or solutions for sudden patchy hair loss (alopecia areata) and possible causes.
2. Determine if the AI generated answer derives information only from the context provided.
3. Check if the answer mentions any specific treatments or solutions for alopecia areata based on the context.
4. Verify that the possible causes mentioned in the answer align with those discussed in the context.

The AI generated answer adheres to the metric as it mentions topical treatments (minoxidil and corticosteroids) for alopecia areata, which is information derived directly from the context. The answer also correctly identifies alopecia areata as a sudden patchy hair loss condition and discusses possible causes (autoimmune disorder, genetics, environmental triggers, stress, and vitamin deficiencies), all of which are mentioned in the context.

Rating: 5 - The metric is followed completely.

 Steps to

###**Observations:**
For answer 3, the groundedness rating is 5 for following all the metrics but got a rating of 4 for the relevance. The reasoning for any point deduction is unclear, as the answer appears fully derived from the context without missing information.


### Query 4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
ground,rel = generate_ground_relevance_response(user_input="What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?",max_tokens=370) #Complete the code to pass the query #4 along with parameters if needed

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the key information related to treatments for brain injuries mentioned in the context.
2. Determine if the AI generated answer only includes information derived from the context.
3. Evaluate the extent to which the metric is followed.

Explanation:
The AI generated answer mentions that rehabilitation is needed when neurologic deficits persist due to brain injury and this rehabilitation should be provided through a team approach that includes physical, occupational, and speech therapy, skill-building activities, and counseling to meet the patient's social and emotional needs. For patients with prolonged comas, a prolonged period of rehabilitation, particularly in cognitive and emotional areas, is often required.

The answer only includes information derived from the context as it directly quotes the text and expands upon the information provided in the context regarding the need for rehabilitation and the components of the rehabilitation proces

###**Observations:**
* For answer 4, The groundedness score is 5 since the response is derived solely from the context.The relevance score is 5 as the metric is fully followed, but there is not enough justification.

* One solution is to adjust the relevance prompt to require a brief explanation for the scores given.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
ground,rel = generate_ground_relevance_response(user_input="What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?",max_tokens=370) #Complete the code to pass the query #5 along with parameters if needed

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Steps to evaluate the answer:
1. Identify the information in the context related to precautions and treatment steps for a person with a fractured leg.
2. Compare the information identified in step 1 with the answer provided by the AI system.
3. Determine if the answer is derived only from the information presented in the context.

Explanation:
The answer adheres to the metric as it mentions all the necessary precautions and treatment steps for a person who has fractured their leg, which are directly derived from the context. The context discusses the importance of sterile wound dressings, tetanus prophylaxis, broad-spectrum antibiotics, immobilization using splints or slings, elevation above heart level, and early mobilization for rapidly healing injuries. The answer includes all these points, making it a complete match with the context.

Rating:
Based on the evaluation criteria, I would rate this answer as a 5 because the metric is followed completely.

 To evaluate the context as pe

###**Observations**
Again for answer 5, the rating for groundedness is 5 and relevance is 4. As per the evaluation the answer cover all the important aspect of the question and provides sufficient information.While the answer is comprehensive, there is an opportunity to enhance its conciseness and specificity.

## Actionable Insights and Business Recommendations

## **Insights:**
* RAG-model responses are clear, trustworthy, and relevant, improving user experience and trust.

* Continuous parameter tuning and prompt adjustments can further improve the responses.

* The RAG model offers the potential for automating medical Q&As, thereby reducing expenses associated with consultations and telemedicine support services.

##**Business Recommendations:**
* Fine-tune the model with advanced embedding models and sentence transformers specialized for medical and healthcare data to enhance the accuracy of the responses.

* Hire prompt engineering specialists to refine prompts for improved relevance and accuracy of the answers.

* The RAG model should be updated and fine tuned continuously to adhere to the lastest medical guidelines and compliance standards.

* The RAG model can be trained on department specific medical data to improve accuracy and relevance for particular medical conditions.




<font size=6 color='blue'>Power Ahead</font>
___